### PR TITLE
Reimagine the tour of C#

### LINKED ISSUE #38572: Quick resources for Developers that know other languages
Readers with experience in other languages want to find quick information about how to learn C# (and .NET) from the perspective of  the language they already know. The first resources should be aimed at developers that know:

- Java
- JavaScript
- Python

---
[Associated WorkItem - 198461](https://dev.azure.com/msft-skilling/Content/_workitems/edit/198461)

### LINKED ISSUE #39813: Rewrite and modernize the Tour of C#
### Type of issue

Outdated article

### Description

The text from this came from the original Microsoft C# 1.0 specification. It's been updated in small ways since then, but it doesn't reflect today's C# features. 

See #38572, which might be related. The tour of C# would be a good location for those introductory articles.

### Page URL

https://learn.microsoft.com/en-us/dotnet/csharp/tour-of-csharp/

### Content source URL

https://github.com/dotnet/docs/blob/main/docs/csharp/tour-of-csharp/index.md

### Document Version Independent Id

ea7fbbaf-fd41-75cb-28fb-a2f36c8d3967

### Article author

@BillWagner

### Metadata

* ID: 14618620-83fc-fd68-93a9-03626c76d680 
* Service: **dotnet-csharp**
* Sub-service: **get-started**

---
[Associated WorkItem - 226921](https://dev.azure.com/msft-skilling/Content/_workitems/edit/226921)

### PR DESCRIPTION
**Pull request by BillWagner**
The old tour was accurate, but way too long, and emphasized too many older concepts.

Make fresh.

FIxes #39813 
Fixes #38572 

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/csharp/methods.md](https://github.com/dotnet/docs/blob/72931c54a61175b78ffc9debec55381281140286/docs/csharp/methods.md) | [Methods in C\#](https://review.learn.microsoft.com/en-us/dotnet/csharp/methods?branch=pr-en-us-40082) |
| [docs/csharp/tour-of-csharp/index.md](https://github.com/dotnet/docs/blob/72931c54a61175b78ffc9debec55381281140286/docs/csharp/tour-of-csharp/index.md) | [A tour of C# - Overview](https://review.learn.microsoft.com/en-us/dotnet/csharp/tour-of-csharp/index?branch=pr-en-us-40082) |
| [docs/csharp/tour-of-csharp/tips-for-java-developers.md](https://github.com/dotnet/docs/blob/72931c54a61175b78ffc9debec55381281140286/docs/csharp/tour-of-csharp/tips-for-java-developers.md) | [Roadmap for Java developers learning C\#](https://review.learn.microsoft.com/en-us/dotnet/csharp/tour-of-csharp/tips-for-java-developers?branch=pr-en-us-40082) |
| [docs/csharp/tour-of-csharp/tips-for-javascript-developers.md](https://github.com/dotnet/docs/blob/72931c54a61175b78ffc9debec55381281140286/docs/csharp/tour-of-csharp/tips-for-javascript-developers.md) | [Roadmap for JavaScript and TypeScript developers learning C\#](https://review.learn.microsoft.com/en-us/dotnet/csharp/tour-of-csharp/tips-for-javascript-developers?branch=pr-en-us-40082) |
| [docs/csharp/tour-of-csharp/tips-for-python-developers.md](https://github.com/dotnet/docs/blob/72931c54a61175b78ffc9debec55381281140286/docs/csharp/tour-of-csharp/tips-for-python-developers.md) | [docs/csharp/tour-of-csharp/tips-for-python-developers](https://review.learn.microsoft.com/en-us/dotnet/csharp/tour-of-csharp/tips-for-python-developers?branch=pr-en-us-40082) |

### CODE REVIEW

File: docs/csharp/tour-of-csharp/tips-for-javascript-developers.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+1. [***Nullable and non-nullable types***](../nullable-references.md): C# supports *nullable value types*, and *nullable reference types* by appending the `?` suffix to a type. For nullable types, the compiler warns you if you haven't checked for `null` before dereferencing the expression. For non-nullable types, the compiler warns you if you might be assigning a `null` value to that variable. These features can minimize your application throwing a <xref:System.NullReferenceException?displayProperty=nameWithType>. The syntax may be familiar from TypeScript's use of `?` for optional properties.
+1. [***Properties***](../properties.md) and [***Indexers***](../indexers.md): Both properties and indexers (treating a class like an array or dictionary) have language support. In Java, they are naming conventions for methods starting with `get` and `set`.
```
Comment (diberry) [inline docs/csharp/tour-of-csharp/tips-for-javascript-developers.md:24]:
```suggestion
1. [***Properties***](../properties.md) and [***Indexers***](../indexers.md): Both properties and indexers (treating a class like an array or dictionary) have language support. In JavaScript, they are naming conventions for methods starting with `get` and `set`.
```

File: docs/csharp/tour-of-csharp/tips-for-javascript-developers.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+1. [***Extension methods***](../programming-guide/classes-and-structs/extension-methods.md):  In C#, you can create methods that *extend* a class or interface. These enable you to extend the behavior of a type from a library, or all types that implement a given interface.
+1. [***LINQ***](../linq/index.md): Language integrated query (LINQ) provides a common syntax to query and transform data, regardless of its storage.
+
+As you learn more, you'll find other differences as well, but many of those are smaller in scope.
+
+You will find familiar features and idioms from JavaScript and TypeScript that aren't available in C#:
+
+1. ***dynamic types***: C# uses static typing. A variable has a type when it is declared, and that type can't change.
+1. ***Prototypal inheritance***: C# inheritance is part of the type declaration. A C# `class` declaration states any base class. In JavaScript, you can set the `__proto__` property to set the base type on any instance.
+1. ***Interpreted language***: C# code must be compiled before it's run. JavaScript and TypeScript code can be run directly in the browswer.
```
Comment (diberry) [inline docs/csharp/tour-of-csharp/tips-for-javascript-developers.md:34]:
```suggestion
1. ***Interpreted language***: C# code must be compiled before it's run. JavaScript and TypeScript code can be run directly in the browser.
```

Comment (diberry) [reply docs/csharp/tour-of-csharp/tips-for-javascript-developers.md:34]: I'm not sure I've seen Type run directly in the browser. How is that done?

File: docs/csharp/tour-of-csharp/tips-for-javascript-developers.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+In addition, a few TypeScript features aren't available in C#:
+
+1. ***Union types***: C# doesn't support union types. However, design proposals are in progress.
+1. ***Decorators***: C# doesn't have decorators. Some common decorators, such as `@sealed` are reserved keywords in C#. Others may be implemented using [Attributes](../language-reference/attributes/general.md).
```
Comment (diberry) [inline docs/csharp/tour-of-csharp/tips-for-javascript-developers.md:39]: I've always thought of decorators as the JS version of attributes.

**Review by diberry — CHANGES_REQUESTED** [review]
Looks good. Thank you for sharing.

**Comment by richlander** [comment]
Not specific to the edits, but the existing `methods.md` doc:

> This topic discusses named methods. For information about anonymous functions, see [Lambda expressions](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/lambda-expressions).

That's a lot of complexity to start with (methods vs functions, named vs anonymous, ...) and I don't think relevant enough. It does include the more relevant concept of expression-bodied members at the end (which is fine). However, there is no obvious linkage between this call out and that more relevant content.

```csharp
namespace MotorCycleExample
{
    abstract class Motorcycle
    {
        // Anyone can call this.
        public void StartEngine() {/* Method statements here */ }

        // Only derived classes can call this.
        protected void AddGas(int gallons) { /* Method statements here */ }

        // Derived classes can override the base class implementation.
        public virtual int Drive(int miles, int speed) { /* Method statements here */ return 1; }

        // Derived classes can override the base class implementation.
        public virtual int Drive(TimeSpan time, int speed) { /* Method statements here */ return 0; }

        // Derived classes must implement this.
        public abstract double GetTopSpeed();
    }
```

This example uses the old syntax style and an `abstract` class, which I think is an advanced topic. I think a simpler example with modern syntax would be much preferred. I grasp that the example is used later, just don't start with it. Also, aren't interfaces a lot more common than abstract classes?

> class TestMotorcycle : Motorcycle

It's very strange that we start with an inheritance example before the inheritance section and then don't use example. I'd put all the inheritance discussion at the end. The other sections seem more important.

File: docs/csharp/tour-of-csharp/tips-for-javascript-developers.md (hunk 2 above)
Comment (BillWagner) [reply docs/csharp/tour-of-csharp/tips-for-javascript-developers.md]: > I'm not sure I've seen TypeScript run directly in the browser. How is that done?

You can load the TypeScript compiler into the browser, and compile it. It's how the [TypeScript playground](https://www.typescriptlang.org/play) does it. For this article, I removed the note on TypeScript, as that's not really a common scenario.

File: docs/csharp/tour-of-csharp/tips-for-python-developers.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+1. [***Static typing***](../fundamentals/types/index.md): In C#, a variable declaration includes its type. Reassigning a variable to an object of a different type generates a compiler error. In python, the type can change when reassigned.
+1. [***Nullable types***](../nullable-references.md): C# variables can be *nullable* or *non-nullable*. A non-nullable type is one that can't be null (or nothing). It always refers to a valid object. By contrast, a nullable type might either refer to a valid object, or null.
+1. [***LINQ***](../linq/index.md): The query expression keywords that make up Language Integrated Query (LINQ) aren't keywords in Python. However, Python libraries like `itertools`, `more-itertools`, and `py-linq` provide similar functionality.
+1. [***Generics***](../fundamentals/types/generics.md): C# generics use C# static typing to make assertions about the arguments supplied for type parameters. A generic algorithm might need to specify constraints that an argument type must satisfy.
+
+Finally, there are some features of Python aren't available in C#:
```
Comment (MadsTorgersen) [inline docs/csharp/tour-of-csharp/tips-for-python-developers.md:26]:
```suggestion
Finally, there are some features of Python that aren't available in C#:
```

**Review by MadsTorgersen — APPROVED** [review]
I really like this rewrite! A modernization was needed, and this does a really good job, both updating the focus and removing detail. I love the tone of the sections for people coming from other languages - it really reinforces a "can-do" approach, helping new C# learners build on their strengths and not be daunted by differences.

File: docs/csharp/tour-of-csharp/index.md [hunk 5]
```diff
@@ -1,148 +1,79 @@
...
+The preceding query works for many storage types represented by `Students`. It could be a collection of objects, a database table, a cloud storage blob, or an XML structure. The same query syntax works for all storage types.
 
-As a small example, consider the below C# code:
+The [Task based asynchronous programming model (TAP)](../asynchronous-programming/index.md) enables you to write code that reads as though it runs synchronously, even though it runs asynchronously. It utilizes the `async` and `await` keywords to describe methods that are asynchronous, and when an expression evaluates asynchronously. The following sample awaits an asynchronous web request. When the asynchronous operation completes, the method returns the length of the response:
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/tour-of-csharp/index.md:59]: I've only seen it in the titles of the whitepapers. Removing it.

**Comment by BillWagner** [comment]
closing and reopening for a fresh CI build

**Pull request merged**